Add uninstall script and make the bootstrap script install-only

matteodelabre opened the pull request:
This PR adds a new `toltecctl` script, installed by the `toltec-bootstrap` package, that supersedes the former `entware-reenable` script and provides a way to uninstall Toltec (fixes #295). This script could be extended to add new install management features in the future. The uninstall script does the following things:

* Uninstalls all Toltec and Entware packages, running removal scripts for each package.
* Removes the bind mount of Toltec files on /opt (the mount is removed lazily so that it does not fail even if some processes that have files opened in that folder are still running)
* Removes all Toltec files
* Clears Toltec definitions from .bashrc
* Re-enables xochitl
* Reboots the device

Another major change of this PR is to make the bootstrap script fail if files from previous installs remain in the folders used by Toltec (fixes #332). With this change, the bootstrap script only has a single role, that is to create a new Toltec install in a clean system. Roles previously assumed explicitly or implicitly by the bootstrap script are deprecated:

* To re-enable an existing install after a system update, users should run `toltecctl enable` instead of re-running the bootstrap script.
* To create a new clean install, users should first uninstall the existing one and then run the bootstrap script.
* To switch to another branch (see #341), users should manually edit the `/opt/etc/opkg.conf` file instead of re-running the bootstrap script (a future PR could extend the `toltecctl` script to handle switching branches).

Test plan
=========

(Only tested on rM2 so far.)

Uninstall script
----------------

* Upgrade the toltec-bootstrap package to version 0.0.5-1 (this will remove the `bootstrap` and `entware-reenable` scripts from the home folder and add the `toltecctl` script to the `~/.local/bin` folder)
* Run `~/.local/bin/toltecctl enable` and re-source .bashrc (to add .local/bin to the PATH)
* Run `toltecctl uninstall`
* After the tablet reboots, make sure no traces of the Toltec install remain (in ~/.entware, /opt, /lib/systemd/system/opt.mount)
* If you had disabled xochitl, make sure that it starts up again after uninstalling

Re-enable script
----------------

* Upgrade the toltec-bootstrap package to version 0.0.5-1 (this will remove the `bootstrap` and `entware-reenable` scripts from the home folder and add the `toltecctl` script to the `~/.local/bin` folder)
* Simulate a system update, for example by switching to the other root partition or using [remarkable-update](https://github.com/ddvk/remarkable-update)
* Run `~/.local/bin/toltecctl enable`
* Make sure that Toltec has been re-enabled (files visible in /opt, opkg command working, apps running)

Bootstrap script
----------------

* Copy the new bootstrap script to the tablet.
* Try running the bootstrap script with an existing Toltec install (the script should fail and tell you that Toltec is already installed)
* Try running the bootstrap script with only some parts of a Toltec install remaining (for example, only the .entware folder - the script should fail and tell you which files need to be cleaned up)
* Run the bootstrap script after uninstalling Toltec or from a clean system, and make sure the install works.

matteodelabre committed:
Add uninstall script and make the bootstrap script install-only

This PR adds a new `toltecctl` script, installed by the `toltec-bootstrap` package, that supersedes the former `entware-reenable` script and provides a way to uninstall Toltec (fixes #295). This script could be extended to add new install management features in the future. The uninstall script does the following things:

* Uninstalls all Toltec and Entware packages, running removal scripts for each package.
* Removes the bind mount of Toltec files on /opt (the mount is removed lazily so that it does not fail even if some processes that have files opened in that folder are still running)
* Removes all Toltec files
* Clears Toltec definitions from .bashrc
* Re-enables xochitl
* Reboots the device

Another major change of this PR is to make the bootstrap script fail if files from previous installs remain in the folders used by Toltec (fixes #332). With this change, the bootstrap script only has a single role, that is to create a new Toltec install in a clean system. Roles previously assumed explicitly or implicitly by the bootstrap script are deprecated:

* To re-enable an existing install after a system update, users should run `toltecctl enable` instead of re-running the bootstrap script.
* To create a new clean install, users should first uninstall the existing one and then run the bootstrap script.
* To switch to another branch (see #341), users should manually edit the `/opt/etc/opkg.conf` file instead of re-running the bootstrap script (a future PR could extend the `toltecctl` script to handle switching branches).

Test plan
=========

(Only tested on rM2 so far.)

Uninstall script
----------------

* Upgrade the toltec-bootstrap package to version 0.0.5-1 (this will remove the `bootstrap` and `entware-reenable` scripts from the home folder and add the `toltecctl` script to the `~/.local/bin` folder)
* Run `~/.local/bin/toltecctl enable` and re-source .bashrc (to add .local/bin to the PATH)
* Run `toltecctl uninstall`
* After the tablet reboots, make sure no traces of the Toltec install remain (in ~/.entware, /opt, /lib/systemd/system/opt.mount)
* If you had disabled xochitl, make sure that it starts up again after uninstalling

Re-enable script
----------------

* Upgrade the toltec-bootstrap package to version 0.0.5-1 (this will remove the `bootstrap` and `entware-reenable` scripts from the home folder and add the `toltecctl` script to the `~/.local/bin` folder)
* Simulate a system update, for example by switching to the other root partition or using [remarkable-update](https://github.com/ddvk/remarkable-update)
* Run `~/.local/bin/toltecctl enable`
* Make sure that Toltec has been re-enabled (files visible in /opt, opkg command working, apps running)

Bootstrap script
----------------

* Copy the new bootstrap script to the tablet.
* Try running the bootstrap script with an existing Toltec install (the script should fail and tell you that Toltec is already installed)
* Try running the bootstrap script with only some parts of a Toltec install remaining (for example, only the .entware folder - the script should fail and tell you which files need to be cleaned up)
* Run the bootstrap script after uninstalling Toltec or from a clean system, and make sure the install works.

matteodelabre committed:
Fix formatting

matteodelabre committed:
Fix bootstrap checksum

Eeems commented on package/toltec-bootstrap/package:
```diff
@@ -3,24 +3,17 @@
 url=https://toltec-dev.org/
-pkgver=0.0.4-1
-timestamp=2021-04-04T21:55Z
+pkgver=0.0.5-1
```
```suggestion
pkgver=0.0.6-1
```
There is another PR with the version being bumped to 0.0.5-1

LinusCDE committed:
[plato] Update to 0.9.17-1 (#344)

* Update plato to 0.9.16-1

* Fix formatting

* Fix linting errors

* Update messages for metadata updates

* Update plato to 0.9.17

* Improve migration script

* Do not print/do anything if there’s no .metadata.json file
* Do the migration even if there’s already a .old file, by creating a unique backup file name
* Tell the user where the backup is stored

Co-authored-by: raisjn <70462544+raisjn@users.noreply.github.com>
Co-authored-by: Mattéo Delabre <spam@delab.re>
Co-authored-by: Mattéo Delabre <1370040+matteodelabre@users.noreply.github.com>

plan5 committed:
[innernet] wrapper for wireguard (#350)

* Add innernet-client package

Co-authored-by: plan5 <plan5 30434574+plan5@users.noreply.github.com>
Co-authored-by: Nathaniel van Diepen <Eeems@users.noreply.github.com>
Co-authored-by: Mattéo Delabre <1370040+matteodelabre@users.noreply.github.com>
Co-authored-by: Mattéo Delabre <spam@delab.re>

plan5 committed:
[gocryptfs] Add gocryptfs package (build without openssl) (#318)

*Add gocryptfs package

Co-authored-by: plan5 <plan5 30434574+plan5@users.noreply.github.com>
Co-authored-by: Nathaniel van Diepen <Eeems@users.noreply.github.com>
Co-authored-by: Mattéo Delabre <1370040+matteodelabre@users.noreply.github.com>

LinusCDE reviewed (changes requested):
Tested most of your steps on both devices and they seem to be fine for the most part.

For bootstrap I used `toltec_branch=testing` since the architecture urls don't exist on stable yet.

A cool feature would be to have toltec-bootstrap automatically add `/home/root/.local/bin` to `PATH` upon upgrade to the newer version.

Disabling toltec also doesn't affect the services. It removes `/opt`, but systemd may still be configured to attempt launching remux. In my case it also didn't fall back to xochitl, so resetting the service files similar to on uninstall may be needed. Or at least warn the user about what to do.

TL;DR:
 - Also add new path to .bashrc upon package update
 - Warn user or manually force xochitl to be enabled on `toltecctl disable`

This is a really cool utility. It'll definitely make toltec more user friendly! :+1:

Eeems committed:
Update innernet to v1.3.0 (#358)

* Update to innernet 1.3.0

Eeems committed:
Update KOReader to v2021.05 (#357)

matteodelabre committed:
Update VNSee to 0.4.0 and add launcher files (#355)

- Add reMarkable 2 support (depends on [remarkable2-framebuffer](https://github.com/ddvk/remarkable2-framebuffer)).
- Tolerate server resolutions differing from 1404x1872 or 1404x1872.
    - The server image is cropped to fit in the screen.
- Add `vnsee-gui` script to start VNSee from a GUI.
    - This is a [simple](https://rmkit.dev/apps/sas) script.

This PR additionally contains launcher files for Draft, remux, and Oxide
that start the `vnsee-gui` script.

Co-authored-by: Linus K <linus@cosmos-ink.net>

LinusCDE committed:
Add bash-completion and toltec-completion packages (#277)

* Add bash-completion and toltec-completion packages

* Fix format issues

* Small fixes and update copyright year

* Revert "Fix format issues"

This reverts commit 83e26b63551fd4c291d0a894266277655795ff6f.

* Fix depends keyword in package file

* Fix some shellcheck issues and ignore others

Reasoning was added to the top of the file.

* Remove source hint from toltec-completion

* Compromize more to make shfmt happy

Seems that shfmt complains even after I disabled
the shellformat rule. Does shfmt not use shellformat?

* Ignore another check

The completions should not be suspect to the problems
listed here: https://www.shellcheck.net/wiki/SC2207

* Add completions for toltecctl

The command is still a PR (#356).

Co-authored-by: Nathaniel van Diepen <Eeems@users.noreply.github.com>

LinusCDE committed:
Update chessmarkable to 0.7.0-1 (#362)

Co-authored-by: Nathaniel van Diepen <Eeems@users.noreply.github.com>

matteodelabre committed:
Update package/toltec-bootstrap/package

Co-authored-by: Nathaniel van Diepen <Eeems@users.noreply.github.com>

matteodelabre committed:
Add confirmation prompt before uninstalling

matteodelabre committed:
Merge branch 'testing' into package/toltec-bootstrap/add-toltecctl

matteodelabre committed:
Fix formatting

matteodelabre committed:
Remove option to disable and rename 'enable' to 'reenable'

matteodelabre committed:
Remove old function

matteodelabre committed:
Add 'toltecctl set-path' to auto-update bashrc & call it on upgrade

matteodelabre commented:
Thank you very much for your review @LinusCDE!

> A cool feature would be to have toltec-bootstrap automatically add `/home/root/.local/bin` to `PATH` upon upgrade to the newer version.

I added a new `toltecctl set-path` action that automatically fixes or upgrades the PATH in bashrc. I added this command to the package’s `configure()` block so that it is invoked when the package is updated.

> Disabling toltec also doesn't affect the services. It removes `/opt`, but systemd may still be configured to attempt launching remux. In my case it also didn't fall back to xochitl, so resetting the service files similar to on uninstall may be needed. Or at least warn the user about what to do.

This is a good point. Actually, I’m not sure whether the action of disabling Toltec without uninstalling it is useful to have in `toltecctl`. I assume most users will want to completely uninstall it. So I’ve removed this option to make things less confusing.

Also, thanks for working on the Bash completions! I updated the `_toltecctl` completion script to match the new set of actions.

LinusCDE commented:
Cool. I tested the `set-path` command on the rm1 and it works fine!

Would you want me to retest everything again or can I just accept it? I looked at the new commits and I don't see any issues or things that need a full re-review imo.

matteodelabre commented:
> Would you want me to retest everything again or can I just accept it? I looked at the new commits and I don't see any issues or things that need a full re-review imo.

On my side, I re-did the tests for the uninstall and bootstrap scripts (on rM2), and also tested that upgrading the package sets the PATH as expected. I don’t think it’s necessary to re-do all the other tests since there weren’t any major changes.

matteodelabre commented:
Maybe we should revert @Eeems’ suggestion from above of bumping `toltec-reenable`’s version from 0.0.4 to 0.0.6 if this PR is to be merged before #342.

matteodelabre committed:
Revert toltec-bootstrap version to 0.0.5

matteodelabre commented:
Reverted the version to 0.0.5. I will set PR #342’s version to 0.0.6.

isaacwisdom commented:
Hi, I'm trying to run the uninstall command added, but get this error:
"/home/root/.local/bin/toltecctl: line 181: uninstall_force: command not found"
Any ideas what's going on?

isaacwisdom commented:
> Hi, I'm trying to run the uninstall command added, but get this error:
> "/home/root/.local/bin/toltecctl: line 181: uninstall_force: command not found"
> Any ideas what's going on?

Just kidding, figured it out. uninstall-force() is defined but uninstall() references uninstall_force(). Changing the underline  to a hyphen fixes the issue.
